algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+variables (T' : set (α/S)) [is_ideal (α/S) T']
+
+@[reducible] def ideal_to_quotient : set (α/S) := is_ideal.to_coset S '' T
+@[reducible] def quotient_to_ideal : set α := is_ideal.to_coset S ⁻¹' T'
+
+instance ideal_to_quotient.is_ideal : is_ideal (α/S) (ideal_to_quotient S T) :=
+{ zero_mem := ⟨0, is_ideal.zero_mem T, rfl⟩,
+  add_mem  := λ x y ⟨m, hm1, hm2⟩ ⟨n, hn1, hn2⟩,
+    ⟨m + n, is_ideal.add_mem hm1 hn1, by rw [← hm2, ← hn2]; refl⟩,
+  mul_mem  := λ x y ⟨m, hm1, hm2⟩,
+    let ⟨n, hn⟩ := is_ideal.coset_rep y in
+      ⟨m * n, is_ideal.mul_mem hm1, by rw [← hm2, ← hn]; refl⟩ }
+
+def quotient_to_ideal.is_ideal : is_ideal α (quotient_to_ideal S T') :=
+is_ideal.hom_preimage (is_ideal.to_coset S) T'
+
+theorem quotient_to_ideal.contains : S ⊆ quotient_to_ideal S T' :=
+λ x hx, calc ⟦x⟧ = 0 : is_ideal.coset_zero_of_mem S x hx
+             ... ∈ T' : is_ideal.zero_mem T'
+
+theorem contains_to_quotient_to_contains : S ⊆ T → quotient_to_ideal S (ideal_to_quotient S T) = T :=
+λ h, set.ext $ λ x,
+  ⟨λ ⟨z, hz1, hz2⟩, by simpa using (is_ideal.sub_mem hz1 $ h (by simpa using hz2 : z - x ∈ S)),
+   λ hx, ⟨x, hx, rfl⟩⟩
+
+theorem quotient_to_contains_to_quotient : ideal_to_quotient S (quotient_to_ideal S T') = T' :=
+set.ext $ λ x,
+  ⟨λ ⟨z, hz1, hz2⟩, by rwa ← hz2,
+   λ hx, let ⟨z, hz⟩ := is_ideal.coset_rep x in ⟨z, by rwa ← hz at hx, hz⟩⟩
+
+theorem ideal_to_quotient.subset : T₁ ⊆ T₂ → ideal_to_quotient S T₁ ⊆ ideal_to_quotient S T₂ :=
+λ h z ⟨w, hw1, hw2⟩, ⟨w, h hw1, hw2⟩
+
+theorem ideal_to_quotient.univ : ideal_to_quotient S (univ_ideal α) = univ_ideal (α/S) :=
+set.ext $ λ x, ⟨λ hx, trivial, λ hx, by simpa using is_ideal.coset_rep x⟩
+
+theorem quotient_to_ideal.subset (T₁ : set (α/S)) [is_ideal (α/S) T₁] (T₂ : set (α/S)) [is_ideal (α/S) T₂] :
+  T₁ ⊆ T₂ → quotient_to_ideal S T₁ ⊆ quotient_to_ideal S T₂ :=
+λ h z hz, h hz
+
+theorem quotient_to_ideal.zero : quotient_to_ideal S (zero_ideal $ α/S) = S :=
+set.ext $ λ x, by simp [is_ideal.coset_zero_iff_mem S x]
+
+theorem quotient_to_ideal.univ : quotient_to_ideal S (univ_ideal $ α/S) = univ_ideal α :=
+set.ext $ λ x, ⟨λ hx, trivial, λ hx, trivial⟩
+
+end prop_1_1
+
+-- Proposition 1.1 end
+
+namespace is_hom
+
+variables {α : Type u} {β : Type v} [comm_ring α] [comm_ring β] (f : α → β) [is_hom f]
+
+@[reducible] def ker : set α := f⁻¹' (zero_ideal β)
+instance ker.is_ideal : is_ideal α (ker f) := is_ideal.hom_preimage f $ zero_ideal β
+
+@[reducible] def im : set β := { y | ∃ x, f x = y }
+instance im.subring : subring β (im f) :=
+{ add_mem := λ x y ⟨m, hm⟩ ⟨n, hn⟩, ⟨m + n, by simp [map_add f, hm, hn]⟩,
+  neg_mem := λ x ⟨m, hm⟩, ⟨-m, by simp [map_neg f, hm]⟩,
+  mul_mem := λ x y ⟨m, hm⟩ ⟨n, hn⟩, ⟨m * n, by simp [map_mul f, hm, hn]⟩,
+  one_mem := ⟨(1:α), map_one f⟩ }
+
+instance im.comm_ring : comm_ring (im f) := subring.comm_ring β (im f)
+
+end is_hom
+
+structure isomorphism (α : Type u) (β : Type v) [comm_ring α] [comm_ring β] :=
```

Comment:
define this as an extension of an `equiv` (from `data.equiv`). Even better: just reuse `linear_equiv` from `algebra/liner_algebra/default.lean`.